Firestore supported matchingengine

Majidbadal opened the pull request:
So far the matching engine vector store module only supported GCS as backend storage. I have added support for firestore as it is a nicer option for document storage. The corresponding matching engine notebook example is adjusted accordingly.



#### Who can review?
Tag maintainers/contributors who might be interested:
@dev2049

Majidbadal committed:
add initial support for firestore in matching engine

Majidbadal committed:
Add the fields in firestore db as metadata to the retrieved Document objects

Majidbadal committed:
Adjust comments to include firestore instructions

Majidbadal committed:
Adjust the matchingengine notebook to include firestore

Majidbadal committed:
Add option to provide metadata for adding new documents to firestore

Majidbadal commented:
@dev2049 I am not sure about the missing checks. However, the code should work fine and can be tested with a matching engine instance already deployed. 
thanks.

Majidbadal committed:
Merge branch 'hwchase17:master' into firestore_supported_matchingengine

Majidbadal committed:
Merge branch 'master' into firestore_supported_matchingengine

hwchase17 reviewed (commented):
is thsi better to have as a separate implementation?

they could still share a lot of logic, but it seems like kinda one or the other so maybe better as two vectorstore implementations

Majidbadal commented:
@hwchase17 thanks for the comment. I am not sure about that because the underlying vector store i.e. vertex's matching engine is the same. The only difference is the backend storage for the documents which would be either GCS or Firestore.
What do you think?

baskaryan commented on langchain/vectorstores/matching_engine.py:
```diff
@@ -80,19 +87,24 @@ def __init__(
         try:
             from google.cloud import aiplatform, storage  # noqa: F401
             from google.oauth2 import service_account  # noqa: F401
+            from google.cloud import firestore
```
if someone was using gas and didn't have firestore installed, this would raise an error right?

Majidbadal replied:
@baskaryan Good catch! fixed now.

Majidbadal committed:
check if firestore is installed only if the firestore parameters are supplied

Majidbadal committed:
move google libraries validator to the end of initializer

Majidbadal committed:
Replace embed_documents with embed_query

Replace embed_document() in the similarity_search model with embed_query(). This way the correct method on the underlying embedding model is called so if Hypothetical document embedding is used, it will first call the chain within the embed_query() method and then generating embeddings for the produced hypothetical answer

Majidbadal committed:
add kwargs to enable filtering of vectors

lkuligin commented:
Maybe it's a good idea to separate an implementation of a Matching Engine (that stores vectores and corresponding metadata like ids) and additional mixins that should store a mappings from doc_id to the text chunk.

Majidbadal committed:
resolved merge conflicts by moving matching engine in a new directory

leo-gan commented:
@Majidbadal Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that, ping me and I push this PR for the review. Thanks!
If this PR is not needed anymore, could you, please, let me know or close this PR?

Majidbadal commented:
@leo-gan Hi,
About the last comment, I don't think a separate implementation is necessary as all mixins and metadata can be stored in firestore using the current implementation.
I resolved the last merge conflict. Please point me to any current issues. Thanks!
